gr-uhd/lib/rfnoc_graph_impl.cpp

```diff
@@ -0,0 +1,193 @@
+            }
+            _graph->connect(_tx_streamers.at(src_block_id),
+                            src_block_port,
+                            block_id_t(dst_block_id),
+                            dst_block_port);
+            return;
+        }
+        if (_rx_streamers.count(dst_block_id)) {
+            _graph->connect(src_block_id,
+                            src_block_port,
+                            _rx_streamers.at(dst_block_id),
+                            dst_block_port);
+            return;
+        }
+
+        _graph->connect(block_id_t(src_block_id),
+                        src_block_port,
+                        block_id_t(dst_block_id),
+                        dst_block_port);
+    }
+
+    void connect(const std::string& block1,
+                 const std::string& block2,
+                 bool skip_property_propagation)
+    {
+        connect(block1, 0, block2, 0, skip_property_propagation);
+    }
+
+    void connect(rfnoc_block::sptr src_block,
+                 const size_t src_block_port,
+                 rfnoc_block::sptr dst_block,
+                 const size_t dst_block_port,
+                 const bool skip_property_propagation)
+    {
+        connect(src_block->get_unique_id(),
+                src_block_port,
+                dst_block->get_unique_id(),
+                dst_block_port,
+                skip_property_propagation);
+    }
+
+
+    ::uhd::rx_streamer::sptr create_rx_streamer(const size_t num_ports,
+                                                const stream_args_t& args)
+    {
+        auto streamer = _graph->create_rx_streamer(num_ports, args);
+        const std::string streamer_id =
+            std::dynamic_pointer_cast<::uhd::rfnoc::node_t>(streamer)->get_unique_id();
+        _rx_streamers.insert({ streamer_id, streamer });
+        return streamer;
+    }
+
+    ::uhd::tx_streamer::sptr create_tx_streamer(const size_t num_ports,
+                                                const stream_args_t& args)
+    {
+        auto streamer = _graph->create_tx_streamer(num_ports, args);
+        const std::string streamer_id =
+            std::dynamic_pointer_cast<::uhd::rfnoc::node_t>(streamer)->get_unique_id();
+        _tx_streamers.insert({ streamer_id, streamer });
+        return streamer;
+    }
+
+    void commit()
+    {
+        if (!_commit_called.exchange(true)) {
+            _graph->commit();
+        }
+    }
+
+    std::string get_block_id(const std::string& block_name,
+                             const int device_select,
+                             const int block_select)
+    {
+        std::string block_hint = block_name;
+        if (device_select >= 0) {
+            block_hint = str(boost::format("%d/%s") % device_select % block_hint);
+        }
+        if (block_select >= 0) {
+            block_hint = str(boost::format("%s#%d") % block_name % block_select);
+        }
```

Comment:
Is there ever a case where both `device_select` and `block_select` are `>= 0`? Because if so, it doesn't seem like this would generate the right `device/name#block` string combination in that case.

Reply:
Ouch.

Reply:
:medal_sports: for spotting that just by reading the code.